packages/apollo-server-lambda/src/ApolloServer.ts

```diff
@@ -116,6 +116,17 @@ export class ApolloServer extends ApolloServerBase {
       context: LambdaContext,
       callback: APIGatewayProxyCallback,
     ) => {
```

Comment:
I think it would be really nice if we could actually write the handler as an async function.

Do you think we could write a generic function like

```
function maybeCallbackify(
      asyncHandler: async (event: APIGatewayProxyEvent, context: LambdaContext) => void
): (event: APIGatewayProxyEvent, context: LambdaContext, callback: APIGatewayProxyCallback | undefined) => PromiseOrValue<void> {
  return (event: APIGatewayProxyEvent, context: LambdaContext, callback: APIGatewayProxyCallback | undefined) => {
    if (callback) {
      asyncHandler(event, context).then(r => callback(null, r), e => callback(e));
      return;
    } else {
      return asyncHandler(event, context);
    }
  }
}
```

or something which will let us make the bulk of this handler into an async function?

Let me know if you're interested in trying this; otherwise I might find time at some point to give it a shot.

Reply:
I think that a good idea. Can we do a follow up PR on that as an enhancement? May be create a issue for that and tag me in. If I get a chance, I will implement it. Until then we could merge this to at least moving forward?

Reply:
Doing that will require a major re factor as most logic will be duplicated and we might need to abstract them out.

Reply:
I have been meaning to take time to understand the Lambda integration better anyway so I went and added a commit to your branch that refactored apollo-server-lambda to be mostly an async function.  I also inlined graphqlLambda — this function was part of the Apollo Server v1 API but has only been an intermediate building block in v2, and having it separate felt like it was more trouble than it was worth here.

Reply:
(I hope you don't mind but I've completely taken over this PR.)

Reply:
This is looking good :)